[lldb/Target] Add GetStartSymbol method to DynamicLoader plugins

medismailben opened the pull request:
This patch introduces a new method to the dynamic loader plugin, to fetch its `start` symbol.

This can be useful to resolve the `start` symbol address for instance.

jasonmolenda commented on lldb/include/lldb/Target/DynamicLoader.h:
```diff
@@ -329,6 +330,11 @@ class DynamicLoader : public PluginInterface {
   /// safe to call certain APIs or SPIs.
   virtual bool IsFullyInitialized() { return true; }
 
+  /// Return the `start` function \b symbol in the dynamic loader module.
```
The meaning of this is clear in a Darwin environment, but I wonder if we should explain it a little more, like "This is the symbol the process will begin executing with `process launch --stop-at-entry`".  I don't feel strongly about this.

jasonmolenda commented on lldb/include/lldb/Target/DynamicLoader.h:
```diff
@@ -329,6 +330,11 @@ class DynamicLoader : public PluginInterface {
   virtual bool IsFullyInitialized() { return true; }
 
+  /// Return the `start` function \b symbol in the dynamic loader module.
+  virtual llvm::Expected<lldb_private::Symbol> GetStartSymbol() {
```
Should this be an llvm::Expected?  It's only going to be defined for DynamicLoaderDarwin today.  If some generic code wanted to use this, where it may not be applicable or defined for another DynamicLoader plugin, it'll need to consume the UnimplementedError.  Given the optionality of it, a std::optional seems more natural to me maybe?

medismailben replied:
In my understanding, every dynamic loader should have an equivalent to a start symbol, that's why I went with `Expected` instead of `optional`. I believe that long term, each dynamic loader plugin would have to implement this method for their own platform and only the base class would have the `UnimplementedError`.

jasonmolenda replied:
I would DEFINITELY not bank on this assumption.  The world of environments a debugger needs to support is vast.

bulbazord reviewed (commented):
Looks fine to me, but you probably want @jasonmolenda to sign off to be sure.

medismailben committed:
[lldb/Target] Add GetStartSymbol method to DynamicLoader plugins

This patch introduces a new method to the dynamic loader plugin, to
fetch its `start` symbol.

This can be useful to resolve the `start` symbol address for instance.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

jasonmolenda commented on lldb/include/lldb/Target/DynamicLoader.h:
```diff
@@ -10,9 +10,11 @@
 #include "lldb/Utility/FileSpec.h"
 #include "lldb/Utility/Status.h"
 #include "lldb/Utility/UUID.h"
+#include "lldb/Utility/UnimplementedError.h"
```
We can remove this header file include now.

medismailben replied:
Yup!

jasonmolenda reviewed (approved):
LGTM.

medismailben commented on lldb/include/lldb/Target/DynamicLoader.h:
```diff
@@ -10,9 +10,11 @@
 #include "lldb/Utility/FileSpec.h"
 #include "lldb/Utility/Status.h"
 #include "lldb/Utility/UUID.h"
+#include "lldb/Utility/UnimplementedError.h"
```
```suggestion
```

medismailben committed:
Remove unused header

labath commented:
I'd like to note that, at least in the ELF world, there's no guarantee that there will be an actual *symbol* at the address where the process starts executing. The convention is to call this `_start` but nothing in the system actually relies on that. The ELF header contains the address of the first instruction, the OS jumps to that, and that's it.

So, I think that a more generic interface would be to return an `Address` (which the caller can try to resolve to a symbol if it wants to).

jimingham commented:
I agree with Pavel, we don't for most purposes care whether there's a symbol at the start address (for instance we can use it for the "expression evaluation return breakpoint" as an address.  So we shouldn't require the symbol.

Jim

> On Jul 22, 2024, at 12:04 AM, Pavel Labath ***@***.***> wrote:
> 
> 
> I'd like to note that, at least in the ELF world, there's no guarantee that there will be an actual symbol at the address where the process starts executing. The convention is to call this _start but nothing in the system actually relies on that. The ELF header contains the address of the first instruction, the OS jumps to that, and that's it.
> 
> So, I think that a more generic interface would be to return an Address (which the caller can try to resolve to a symbol if it wants to).
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/llvm/llvm-project/pull/99673#issuecomment-2242235585>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADUPVW7ZFE2ZXN3SJI7ZSMTZNSVILAVCNFSM6AAAAABLE7M55SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDENBSGIZTKNJYGU>.
> You are receiving this because you are on a team that was mentioned.
>

medismailben commented:
@labath @jimingham I've addressed your comment in https://github.com/llvm/llvm-project/pull/99909
